src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/MockGoogleDataprocDAO.scala

```diff
@@ -27,4 +38,30 @@ class MockGoogleDataprocDAO extends DataprocDAO {
       clusters.remove(clusterName)
     Future(())
   }
+
+  override def updateFirewallRule(googleProject: GoogleProject): Future[Unit] = {
+    Future.successful(
+      if (!firewallRules.contains(googleProject)) {
+        firewallRules += googleProject -> dataprocConfig.clusterFirewallRuleName
+      })
+  }
+
+  override def createBucket(googleProject: GoogleProject, bucketName: String): Future[Unit] = {
+    Future.successful(buckets += googleProject -> bucketName)
+  }
+
+  override def uploadToBucket(googleProject: GoogleProject, bucketName: String, fileName: String, content: File): Future[Unit] = {
+    addToBucket(googleProject, bucketName, fileName)
+  }
+
+  override def uploadToBucket(googleProject: GoogleProject, bucketName: String, fileName: String, content: String): Future[Unit] = {
+    addToBucket(googleProject, bucketName, fileName)
+  }
+
+  private def addToBucket(googleProject: GoogleProject, bucketName: String, fileName: String): Future[Unit] = {
+    Future.successful(if (buckets.exists(_ == (googleProject, bucketName))) {
+      bucketObjects += bucketName -> fileName
+    })
+  }
+
```

Comment:
Do the mock dao functions need error handling?

Reply:
your mock is a pretend google with specific behaviour that you define. in this case your mock is a well-behaved google that never forgets everything, so your tests that use this mock are verifying that leo behaves correctly when talking to a working google. you might also want to make a different mock that's a badly-behaved google that forgets things, or returns 500s, in order to test that leo behaves correctly when talking to crappy google.

in practice rather than making a "nice google" mock and a separate "crappy google" mock, we often jam all of this functionality in the same mock. so you might pass in a google project named "doesnotexist" when calling your mock functions and have your mock return 404 conditional on that